### PR TITLE
[Drop-In UI] Use OnBackPressedDispatcher for back press handling

### PR DESCRIPTION
**Pull request by Zayankovsky**

### CODE REVIEW

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/testutil/TestLifecycleOwner.kt [hunk 1]
```diff
@@ -6,7 +6,7 @@ import androidx.lifecycle.LifecycleRegistry
 
 internal class TestLifecycleOwner : LifecycleOwner {
     @Suppress("MemberVisibilityCanBePrivate")
-    val lifecycleRegistry = LifecycleRegistry(this)
+    val lifecycleRegistry = LifecycleRegistry.createUnsafe(this)
```
Comment (kmadsen) [inline libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/testutil/TestLifecycleOwner.kt:9]: Nice!

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt [hunk 2]
```diff
@@ -13,70 +16,69 @@ import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
...
-        }
-    }
-
-    private fun handleBackPress(): Boolean {
-        return when (store.state.value.navigation) {
+    override fun handleOnBackPressed() {
+        when (store.state.value.navigation) {
             NavigationState.FreeDrive -> {
-                false
+                error("The callback should be disabled in Free Drive state")
```
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:38]: Why error here? It seems like it should be a no-op.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt [hunk 3]
```diff
@@ -13,70 +16,69 @@ import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
...
 /**
- * Key listener for the [NavigationView].
+ * Back pressed callback for the [NavigationView].
  *
  * On back pressed will update the navigation state.
  *
  * (FreeDrive) <- (DestinationPreview) <- (RoutePreview) <- (ActiveNavigation)
  *             <- (Arrival)
  */
 @ExperimentalPreviewMapboxNavigationAPI
```
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:30]:
```suggestion
```

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt [hunk 4]
```diff
@@ -858,45 +857,43 @@ class MapboxNavigationViewCustomizedActivity : DrawerActivity() {
...
         }
 
         override fun onArrival() {
-            lastNavigationState = NavigationState.Arrival
+            binding.navigationView.unregisterMapObserver(onMapLongClick)
+            onBackPressCallback.isEnabled = false
         }
+    }
 
-        override fun onBackPressed(): Boolean {
```
Comment (kmadsen) [inline qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:883]: This api also gives you the ability to customize when back press is handled right? Removing it, removes that ability. And now it errors inside when back pressed is handled in the FreeDrive state. 

It seems we could leave this function so there is a way to control the back press

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt [hunk 5]
```diff
@@ -13,70 +16,69 @@ import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
...
+
+        fun attach(
+            componentActivity: ComponentActivity,
+            store: Store,
+            lifecycleOwner: LifecycleOwner,
+        ) {
+            val onBackPressedCallback = DefaultOnBackPressedCallback(store)
+            componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
+            lifecycleOwner.lifecycleScope.launch {
+                try {
```
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:68]:
```suggestion
```

Why is this needed?

Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:68]: Feels wrong to have try and finally when Flow is designed to control all these things
``` kotlin
lifecycleOwner.lifecycleScope.launch {
    lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
        store.select { it.navigation }
            .map { state -> state != NavigationState.FreeDrive }
            .distinctUntilChanged()
            .onCompletion { onBackPressedCallback.remove() }
            .collect { onBackPressedCallback.isEnabled = it }
    }
}
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt (hunk 2 above)
Comment (Zayankovsky) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:38]: It is impossible to end up here, because the callback is disabled in free drive.

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/testutil/TestLifecycleOwner.kt (hunk 1 above)
Comment (Zayankovsky) [reply libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/testutil/TestLifecycleOwner.kt:9]: There is a standard [`TestLifecycleOwner`](https://developer.android.com/reference/kotlin/androidx/lifecycle/testing/TestLifecycleOwner), where I learned this trick, and which can probably be used instead of own implementations (there are 7 of them, by the way).

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt (hunk 5 above)
Comment (Zayankovsky) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:68]: With your suggestion the callback is removed in `inStop`, but I would to remove it in `onDestroy`.

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt (hunk 4 above)
Comment (kmadsen) [reply qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:883]: But how could you override the behavior of `NavigationView`?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt (hunk 5 above)
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:66]: Yeah this was browser coding, point am making is the `try` `finally` and use the Flow operators.

Would this work?

```
lifecycleOwner.lifecycleScope.launch {
    lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
        val onBackPressedCallback = DefaultOnBackPressedCallback(store)
            componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
        store.select { it.navigation }
            .map { state -> state != NavigationState.FreeDrive }
            .distinctUntilChanged()
            .onCompletion { onBackPressedCallback.remove() }
            .collect { onBackPressedCallback.isEnabled = it }
    }
}
```

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt (hunk 4 above)
Comment (kmadsen) [reply qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:883]: Yeah it is easy to add your own back press. I like that.

I'm saying, `onBackPressed(): Boolean` returns a boolean. It allowed you to disable `NavigationView` back press handling.

For example, you can disable the "go back during active guidance"

Comment (kmadsen) [reply qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:883]: OH nevermind I see

This dispatcher is only enabled for the onRoutePreview callback 🤔 
```
        override fun onRoutePreview() {
            binding.navigationView.unregisterMapObserver(onMapLongClick)
            onBackPressCallback.isEnabled = true
        }
```

Hmm. It works, makes me wonder if it is more difficult to manage the state that way. But maybe it's just different

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt (hunk 2 above)
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:36]: Ok then why crash? Is there an actual error?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt [hunk 6]
```diff
@@ -124,16 +124,12 @@ class NavigationView @JvmOverloads constructor(
 
         MapboxNavigationApp.attach(this)
 
+        val componentActivity = context.toComponentActivity()
         attachCreated(
             AnalyticsComponent(),
-            LocationPermissionComponent(context.toComponentActivityRef(), navigationContext.store),
+            LocationPermissionComponent(context.toComponentActivity(), navigationContext.store),
```
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:130]:
```suggestion
            LocationPermissionComponent(componentActivity, navigationContext.store),
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt [hunk 7]
```diff
@@ -13,70 +15,68 @@ import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
...
  * (FreeDrive) <- (DestinationPreview) <- (RoutePreview) <- (ActiveNavigation)
  *             <- (Arrival)
  */
-@ExperimentalPreviewMapboxNavigationAPI
-internal class OnKeyListenerComponent(
+internal class DefaultOnBackPressedCallback private constructor(
     private val store: Store,
-    private val view: View,
-    private val delegateOnKeyListener: View.OnKeyListener? = null
-) : UIComponent() {
```
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:31]: Even though this does not access MapboxNavigation at the moment, I still think it should be a `UIComponent`

Specifically because we're building tools like `attachCreated` or the `ComponentInstaller`.

When we make a new type of object, it needs to be enabled in a different way https://github.com/mapbox/mapbox-navigation-android/pull/6483/files#diff-1e15f303831c62824e88d37a6920faf7248d2396de4f5763f8524b15c855654bR146

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt [hunk 8]
```diff
@@ -124,16 +124,12 @@ class NavigationView @JvmOverloads constructor(
...
             TripSessionComponent(lifecycle, navigationContext.store),
             MapLayoutCoordinator(navigationContext, binding),
-            OnKeyListenerComponent(
```
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:132]: Will you make it a `UIComponent`? That way it can be moved around with the rest of these components

Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:132]:
```
OnBackPressedComponent(componentActivity, navigationContext.store)
```

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt (hunk 4 above)
Comment (abhishek1508) [reply qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:883]: With this custom callback implementation, this example activity is adding it's own `onBackPressedCallback`. However, the `backpressCallback` added here
```
fun attach(
            componentActivity: ComponentActivity,
            store: Store,
            lifecycleOwner: LifecycleOwner,
        ) {
            val onBackPressedCallback = DefaultOnBackPressedCallback(store)
            componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
```
still remains. I am wondering if we need to expose something that would allow users to remove the `DefaultOnBackPressedCallback` 🤔 Otherwise, we will have multiple `onBackPressedCallback` in this specific case:
- `DefaultOnBackPressedCallback` added by `NavigationView`.
- `MyCustomOnBackPressedCallback` added by example activity.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt (hunk 8 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:132]: Agreed 👍🏻

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt (hunk 7 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:31]: 👍🏻

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt (hunk 4 above)
Comment (Zayankovsky) [reply qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:883]: @abhishek1508 only one callback will process the back press event. `OnBackPressedDispatcher` selects an enabled callback that was added last. If there are no enabled callbacks at the moment, then default behavior will happen (e.g. current activity will finish).

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt (hunk 5 above)
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:66]: Resolving cause this is just ideas

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt (hunk 2 above)
Comment (Zayankovsky) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt:36]: If for some reason the callback becomes enabled in Free Drive (impossible with the current implementation), then nothing will happen when user presses the back button. Activity won't finish and Drop-In won't react in any way. This is clearly undesired behavior. But I can log a message instead of crashing.

Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/DefaultOnBackPressedCallback.kt]: I would agree with logging instead of crashing.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ * Back pressed callback for the [NavigationView].
+ *
+ * On back pressed will update the navigation state.
+ *
+ * (FreeDrive) <- (DestinationPreview) <- (RoutePreview) <- (ActiveNavigation)
+ *             <- (Arrival)
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+internal class BackPressedComponent(
+    componentActivity: ComponentActivity,
```
Comment (tomaszrybakiewicz) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:34]: I don't think we need to inject the ComponentActivity here. All we need is `OnBackPressedDispatcher`.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt [hunk 10]
```diff
@@ -141,13 +141,6 @@ abstract class NavigationViewListener {
      */
     open fun onInfoPanelSlide(slideOffset: Float) = Unit
 
-    /**
-     * Called when hardware back button has been pressed.
-     *
-     * @return True if the listener has consumed the event, false otherwise.
-     */
-    open fun onBackPressed(): Boolean = false
```
Comment (tomaszrybakiewicz) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt:149]: Let's include migration instructions in the CHANGELOG entry.

**Review by tomaszrybakiewicz — COMMENTED** [review]
Nicely done. I left a few comments.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                }
+                NavigationState.Arrival -> {
+                    store.dispatch(endNavigation())
+                }
+            }
+        }
+    }
+
+    init {
+        componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:66]: 👍 

Also, instead of adding the callback on init it can be added `onAttached`. This does not need an `init`

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt (hunk 10 above)
Comment (Zayankovsky) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt:149]: I have already mentioned what should be used instead of `onBackPressed`, isn't that enough?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:66]: Right now it seems meaningless because it is attached `attachCreated`. But it is likely that this component will be needed to move to `attachResumed` at some point. And we'll need to update this anyways

[the remark this answers is not in the document]
Comment (Zayankovsky) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:66]: The problem is that `onAttached(MapboxNavigation)` is called too late. It is even called after `onResume` of the activity that hosts this `NavigationView`. So any `OnBackPressedCallback` developer registers in their activity or fragment will have a lower priority that the one added by `NavigationView`. Actually this is the reason why I didn't implement this as `UIComponent` in the first place.

[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:66]: > is called too late. It is even called after `onResume`

This is a potential issue. It would be good if `onAttached` was called when `onCreate` is called. Sounds like the `ViewLifecycleRegistry` will not move to a created state until the lifecycle is resumed. This is also, something that may be addressed.

But that may have reasons to keep it the way it is, so good reason to keep it in the `init`

Can you add a comment to describe some of this? It is likely to be revisited at some point.

```
    // Add the callback before onAttached so downstream developers can register higher priority
    // back press callbacks.
    init {
        componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
```

[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:66]: > Actually this is the reason why I didn't implement this as UIComponent in the first place.

Seems like an issue that would exist before as well. Because we don't have lifecycle hooks for re-attach that are not part of the resume/pause.

By the way, that is the way compose works too. Composables will not be created until they are "resumed"

[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:66]: Honestly this is looking like a good reason to support the `NavigationViewListener.onBackPressed` because then we have better control of the behavior.

The `re-attach` issue will likely come up again because we are supporting `NavigationView` being re-attached.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt (hunk 10 above)
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt:149]: I think it's nice to provide a short code snippet that shows how to patch old usage. I always found it extremely helpful when upgrading the library versions.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt:66]: Scratch what I was saying, the reattach issue would only happen if we move this component outside on the `attachCreated` lifecycle so the comment should be good enough here.

**Review by kmadsen — APPROVED** [review]
Thanks for addressing feedback! 

[Add a comment](https://github.com/mapbox/mapbox-navigation-android/pull/6483/files#r997177160) to the CREATED lifecycle dependency and LGTM

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt (hunk 10 above)
Comment (Zayankovsky) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/navigationview/NavigationViewListener.kt:149]: I added a link to the documentation at https://developer.android.com. It is hard to provide a short migration guide, because the two approaches are not exactly equivalent. Previously we had to do runtime checks in the callback and return false if we don't want to handle the event, but now we have to disable and enable the callback when necessary.

**Review by abhishek1508 — APPROVED** [review]
LGTM. I compiled and run the branch locally and it's working.
